azurerm_postgresql_server scaling with replica in different region results in error

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform (and AzureRM Provider) Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->
```
Terraform v0.14.9
+ provider registry.terraform.io/hashicorp/azurerm v2.55.0

Your version of Terraform is out of date! The latest version
is 0.14.10. You can update by downloading from https://www.terraform.io/downloads.html
```
### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* `azurerm_postgresql_server`

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Two files:

**postgres.tf**
```hcl
resource "azurerm_resource_group" "rg" {
  name     = var.resource_group
  location = var.main_location
}


resource "azurerm_postgresql_server" "postgres_master" {
  name                             = var.main_db_name
  location                         = var.main_location
  resource_group_name              = azurerm_resource_group.rg.name
  sku_name                         = var.sku_name
  administrator_login              = var.administrator_login
  administrator_login_password     = var.administrator_login_password
  version                          = var.postgres_version
  ssl_enforcement_enabled          = true
  ssl_minimal_tls_version_enforced = "TLS1_2"
  backup_retention_days            = var.retention_days
  geo_redundant_backup_enabled     = var.replicas_count != 0 ? true : false
  storage_mb                       = var.storage_mb
  public_network_access_enabled    = var.public_access_enabled
  auto_grow_enabled                = true

  lifecycle {
    ignore_changes = [
      # Autogrow is enabled
      storage_mb,
    ]
  }

}

resource "azurerm_postgresql_server" "postgres_standby" {
  count                            = var.replicas_count
  name                             = "${azurerm_postgresql_server.postgres_master.name}-r-${var.replicas_count}"
  location                         = var.replicas_location
  resource_group_name              = var.resource_group
  sku_name                         = var.sku_name
  version                          = var.postgres_version
  ssl_enforcement_enabled          = true
  ssl_minimal_tls_version_enforced = "TLS1_2"
  storage_mb                       = var.storage_mb
  public_network_access_enabled    = var.public_access_enabled
  create_mode                      = "Replica"
  creation_source_server_id        = azurerm_postgresql_server.postgres_master.id
  auto_grow_enabled                = true

  lifecycle {
    ignore_changes = [
      # Autogrow is enabled
      storage_mb,
    ]
  }

}
```
**var.tf**
```hcl
provider "azurerm" {
  features {}
}

variable "main_db_name" {
    default = "nfpgtst"
}

variable "main_location" {
    default = "westus2"
}
variable "resource_group" {
    default = "pg-test-diff-reg"
}
variable "sku_name" {
    default = "GP_Gen5_4"
}
variable "administrator_login" {
    default = "nilfranadmin"
}
variable "administrator_login_password" {
    default = "superSecure123$"
}
variable "postgres_version" {
    default = 9.6
}
variable "retention_days" {
    default = 7
}
variable "replicas_count" {
    default = 1
}
variable "storage_mb" {
    default = 5120
}
variable "replicas_location" {
    default = "eastus2"
}
variable "public_access_enabled" {
    default = false
    }

```
### Debug Output
https://gist.github.com/NillsF/cda78b2bb2a7cd90ba833bcced9665ae
<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behaviour
- Scale up should not result in an error. 
### Actual Behaviour
- When scaling up, once the replica gets scaled, an error is emitted:
```
Error: updating PostgreSQL Server "nfpgtst-r-1" (Resource Group "pg-test-diff-reg"): postgresql.ServersClient#Update: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="ServiceBusy" Message="Service is temporarily busy and the operation cannot be performed. Please try again later."
```
This can be resolved by running `terraform plan` and `terraform apply` a second time.
<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->
1. In var.tf, set the following sku_name:
```hcl
variable "sku_name" {
    default = "GP_Gen5_4"
}
```
2. `terraform apply`
3. In var.tf, update sku_name to:
```hcl
variable "sku_name" {
    default = "GP_Gen5_8"
}
```
4. `terraform apply` will emit the error (if replica is in another region)



### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in a Azure China/Germany/Government? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

@aristosvo did some great work implementing locking in the following issue/PR:
* #10754
* #10284

This is easily reproduced by introducing a secondary location for one of the replicas in Acceptance Test `TestAccPostgreSQLServer_scaleReplicas`:
```
❯ make acctests SERVICE='postgres' TESTARGS='-run=Replicas'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/postgres -run=Replicas -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/04/24 15:29:08 [DEBUG] not using binary driver name, it's no longer needed
2021/04/24 15:29:09 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccPostgreSQLServer_scaleReplicas
=== PAUSE TestAccPostgreSQLServer_scaleReplicas
=== CONT  TestAccPostgreSQLServer_scaleReplicas
    testing.go:620: Step 3/6 error: Error running apply: exit status 1
        
        Error: updating PostgreSQL Server "acctest-psql-server-210424152910412500-replica2" (Resource Group "acctestRG-psql-210424152910412500"): postgresql.ServersClient#Update: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="ServiceBusy" Message="Service is temporarily busy and the operation cannot be performed. Please try again later."
        
          on terraform_plugin_test.tf line 43, in resource "azurerm_postgresql_server" "replica2":
          43: resource "azurerm_postgresql_server" "replica2" {
        
        
--- FAIL: TestAccPostgreSQLServer_scaleReplicas (1481.86s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/postgres  1485.338s
FAIL
make: *** [acctests] Error 1
```

I'll check the options to solve it, probably something like removing `sku_name` for a secondary.

--- a/azurerm/internal/services/postgres/postgresql_server_resource.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource.go
@@ -605,6 +605,20 @@ func resourcePostgreSQLServerUpdate(d *schema.ResourceData, meta interface{}) er
 	primaryID := id.String()
 	if mode == postgresql.CreateModeReplica {
 		primaryID = d.Get("creation_source_server_id").(string)
+
+		// Wait for possible restarts triggered by scaling primary (and its replicas)
+		log.Printf("[DEBUG] Waiting for PostgreSQL Server %q (Resource Group %q) to become available", id.Name, id.ResourceGroup)
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{string(postgresql.ServerStateInaccessible), "Restarting"},
+			Target:     []string{string(postgresql.ServerStateReady)},
+			Refresh:    postgreSqlStateRefreshFunc(ctx, client, id.ResourceGroup, id.Name),
+			MinTimeout: 15 * time.Second,
+			Timeout:    d.Timeout(schema.TimeoutCreate),
+		}
+
+		if _, err = stateConf.WaitForState(); err != nil {
+			return fmt.Errorf("waiting for PostgreSQL Server %q (Resource Group %q)to become available: %+v", id.Name, id.ResourceGroup, err)
+		}
 	}
 	locks.ByID(primaryID)
 	defer locks.UnlockByID(primaryID)
@@ -629,13 +643,17 @@ func resourcePostgreSQLServerUpdate(d *schema.ResourceData, meta interface{}) er
 				Sku: sku,
 			}
 			for _, replica := range *listReplicas.Value {
-				future, err := client.Update(ctx, id.ResourceGroup, *replica.Name, propertiesReplica)
+				replicaId, err := parse.ServerID(*replica.ID)
 				if err != nil {
-					return fmt.Errorf("upscaling PostgreSQL Server Replica %q (Resource Group %q): %+v", *replica.Name, id.ResourceGroup, err)
+					return fmt.Errorf("parsing Postgres Server Replica ID : %v", err)
+				}
+				future, err := client.Update(ctx, replicaId.ResourceGroup, replicaId.Name, propertiesReplica)
+				if err != nil {
+					return fmt.Errorf("upscaling PostgreSQL Server Replica %q (Resource Group %q): %+v", replicaId.Name, replicaId.ResourceGroup, err)
 				}
 
 				if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
-					return fmt.Errorf("waiting for update of PostgreSQL Server Replica %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+					return fmt.Errorf("waiting for update of PostgreSQL Server Replica %q (Resource Group %q): %+v", replicaId.Name, replicaId.ResourceGroup, err)
 				}
 			}
 		}

--- a/azurerm/internal/services/postgres/postgresql_server_resource_test.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource_test.go
@@ -754,10 +754,10 @@ func (r PostgreSQLServerResource) createReplica(data acceptance.TestData, sku st
 
 resource "azurerm_postgresql_server" "replica" {
   name                = "acctest-psql-server-%[2]d-replica"
-  location            = azurerm_resource_group.test.location
+  location            = "%[3]s"
   resource_group_name = azurerm_resource_group.test.name
 
-  sku_name = "%[3]s"
+  sku_name = "%[4]s"
   version  = "11"
 
   create_mode               = "Replica"
@@ -765,7 +765,7 @@ resource "azurerm_postgresql_server" "replica" {
 
   ssl_enforcement_enabled = true
 }
-`, r.template(data, sku, "11"), data.RandomInteger, sku)
+`, r.template(data, sku, "11"), data.RandomInteger, data.Locations.Secondary, sku)
 }
 
 func (r PostgreSQLServerResource) createReplicas(data acceptance.TestData, sku string) string {
@@ -774,10 +774,10 @@ func (r PostgreSQLServerResource) createReplicas(data acceptance.TestData, sku s
 
 resource "azurerm_postgresql_server" "replica1" {
   name                = "acctest-psql-server-%[2]d-replica1"
-  location            = azurerm_resource_group.test.location
+  location            = "%[3]s"
   resource_group_name = azurerm_resource_group.test.name
 
-  sku_name = "%[3]s"
+  sku_name = "%[4]s"
   version  = "11"
 
   create_mode               = "Replica"
@@ -791,7 +791,7 @@ resource "azurerm_postgresql_server" "replica2" {
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
 
-  sku_name = "%[3]s"
+  sku_name = "%[4]s"
   version  = "11"
 
   create_mode               = "Replica"
@@ -799,7 +799,7 @@ resource "azurerm_postgresql_server" "replica2" {
 
   ssl_enforcement_enabled = true
 }
-`, r.template(data, sku, "11"), data.RandomInteger, sku)
+`, r.template(data, sku, "11"), data.RandomInteger, data.Locations.Secondary, sku)
 }
 
 func (r PostgreSQLServerResource) createPointInTimeRestore(data acceptance.TestData, version, restoreTime string) string {

--- a/azurerm/internal/services/postgres/postgresql_server_resource_test.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource_test.go
@@ -755,7 +755,7 @@ func (r PostgreSQLServerResource) createReplica(data acceptance.TestData, sku st
 resource "azurerm_resource_group" "replica" {
   name     = "acctestRG-psql-%[2]d-replica"
   location = "%[3]s"
-}  
+}
 
 resource "azurerm_postgresql_server" "replica" {
   name                = "acctest-psql-server-%[2]d-replica"

--- a/azurerm/internal/services/postgres/postgresql_server_resource_test.go
+++ b/azurerm/internal/services/postgres/postgresql_server_resource_test.go
@@ -752,10 +752,15 @@ func (r PostgreSQLServerResource) createReplica(data acceptance.TestData, sku st
 	return fmt.Sprintf(`
 %[1]s
 
+resource "azurerm_resource_group" "replica" {
+  name     = "acctestRG-psql-%[2]d-replica"
+  location = "%[3]s"
+}  
+
 resource "azurerm_postgresql_server" "replica" {
   name                = "acctest-psql-server-%[2]d-replica"
   location            = "%[3]s"
-  resource_group_name = azurerm_resource_group.test.name
+  resource_group_name = azurerm_resource_group.replica.name
 
   sku_name = "%[4]s"
   version  = "11"
@@ -772,10 +777,15 @@ func (r PostgreSQLServerResource) createReplicas(data acceptance.TestData, sku s
 	return fmt.Sprintf(`
 %[1]s
 
+resource "azurerm_resource_group" "replica1" {
+  name     = "acctestRG-psql-%[2]d-replica1"
+  location = "%[3]s"
+}
+
 resource "azurerm_postgresql_server" "replica1" {
   name                = "acctest-psql-server-%[2]d-replica1"
   location            = "%[3]s"
-  resource_group_name = azurerm_resource_group.test.name
+  resource_group_name = azurerm_resource_group.replica1.name
 
   sku_name = "%[4]s"
   version  = "11"

